guide_service_errors.md

```diff
@@ -0,0 +1,234 @@
+-   [Step 3: Apply the Service Errors to your Resource](#step-3-apply-the-service-errors-to-your-resource)
+-   [Step 4: Returning Service Errors](#step-4-returning-service-errors)
+-   [Step 5: Enabling Service Error Validation](#step-5-enabling-service-error-validation)
+
+## Background
+
+It's important to communicate to the consumers of a service what types of errors
+may be returned by a particular API. Fortunately, Rest.li allows application developers to
+configure service errors for their resources. This makes APIs clearer by documenting in
+the IDL what sorts of errors should be expected, allowing clients to effectively handle
+and process returned errors. Also, enabling service error validation allows services
+to guarantee which errors a client should expect to encounter.
+
+## Step 1: Define your Service Errors
+
+First, you need to determine which service error codes your service may return, and what
+information will be contained in error responses corresponding with these codes. You'll
+need to define four things:
+
+- The application-specific string code associated with some service error
+- The HTTP status it'll return
+- The string message it'll return
+- The schema defining the format of the error details
+
+For example, say we want to standardize the way errors are returned as a result of violating
+the service's rate-limiting constraint. We would need to pick a string error code such as
+`QUOTA_EXCEEDED`, an HTTP status such as `429`, a string message describing the failure, and
+some schema describing the custom error details sent back to the client.
+
+## Step 2: Write Your Service Error Definition
+
+Write a Java enum implementing the `ServiceError` interface, which will contain a set of
+service error definitions for your service. The enum implementation may expose methods
+for obtaining the HTTP status, the error code, the error message, and the error detail type.
+Here is an example implementation which defines one service error and exposes all four
+accessor methods:
+
+```java
+public enum SampleServiceError implements ServiceError
+{
+  QUOTA_EXCEEDED(HttpStatus.S_429_TOO_MANY_REQUESTS, "You've exceeded your daily request quota.", QuotaDetails.class),
+  INVALID_PERMISSIONS(HttpStatus.S_403_FORBIDDEN, "User does not have valid permissions", null),
+  ILLEGAL_PARAM(HttpStatus.S_400_BAD_REQUEST, "Request parameter cannot be used", null);
+  // Add more service errors here...
+
+  SampleServiceError(HttpStatus status, String message, Class<? extends RecordTemplate> errorDetailType)
+  {
+    _httpStatus = status;
+    _message = message;
+    _errorDetailType = errorDetailType;
+  }
+
+  private final HttpStatus _status;
+  private final String _message;
+  private final Class<? extends RecordTemplate> _errorDetailType;
+
+  /*
+   * Suggested: Define string constants for each of the error codes for easy access.
+   * These codes should match the enum value names.
+   */
+  public class Codes
```

Comment:
Does Java or Rest.li provide a list of Status code/text the developer can re-use? If they want custom status, they can still define them as illustrated here.

Reply:
There are no provided default error codes, but there are examples for reference such as this. Do you think we should add provided default codes?

Reply:
I remember we had this discussion in the past. In `go` for example, they provide this `https://golang.org/src/net/http/status.go` You can create your own, but the standard ones are already provided.

Reply:
I will ask Karthik and Gevorg about this.